### PR TITLE
fix for 1.0.2 crash due to persistent wrong usage

### PR DESCRIPTION
**Pull request by GuuBu**
When adding record to redirection cache hashtable, it also need to take care of the lifecycle of the string added, which should have same lifecycle with the hash table.  Otherwise, it will contains invalid memory type during the request shutdown and module shutdown process.

### CODE REVIEW

**Review by hzhang87 — APPROVED** [review]

**Review by cmb69 — APPROVED** [review]

**Review by hzhang87 — APPROVED** [review]

File: mysqlnd_azure.c
Comment (hzhang87) [inline mysqlnd_azure.c:268]: Just wondering if PHP has functionality of debug trace. If yes, could we add debug trace (or error trace) for negative results or exception situations. You can image that you cannot debug on customer's machine. So, when problem happened, the only thing I can image that we can do is to ask them enable debug trace if we fail to identify the root cause. But you can have your suggestion that how to handle with such case.

File: mysqlnd_azure.c
Comment (hzhang87) [inline mysqlnd_azure.c:676]: Remember to format the code (tab) in the future

**Review by hzhang87 — APPROVED** [review]
Approved to unblock and complete anytime. But please get Dion's approval before complete

File: mysqlnd_azure.c
Comment (cmb69) [reply mysqlnd_azure.c:268]: You could use [zend_fetch_debug_backtrace()](https://github.com/php/php-src/blob/php-7.3.13/Zend/zend_builtin_functions.c#L2456), and print it with [zend_print_zval_r_to_str](https://github.com/php/php-src/blob/php-7.3.13/Zend/zend.c#L487) (or similar), if a respective INI option is set (maybe use `mysqlnd.debug`).

File: redirect_cache.c
Comment (dionysus1016) [inline redirect_cache.c:48]: About the format, I think it maybe editor related. It looks with right indent with my editor...

File: redirect_cache.c
Comment (dionysus1016) [inline redirect_cache.c:73]: code format

File: mysqlnd_azure.c
Comment (GuuBu) [reply mysqlnd_azure.c:676]: About the format, I think it maybe editor related. It looks with right indent with my editor...

File: redirect_cache.c
Comment (GuuBu) [reply redirect_cache.c:73]: About the format, I think it maybe editor related. It looks with right indent with my editor...

File: mysqlnd_azure.c
Comment (GuuBu) [reply mysqlnd_azure.c:268]: Thanks, I think we can do it.

**Review by dionysus1016 — APPROVED** [review]

File: redirect_cache.c
Comment (cmb69) [reply redirect_cache.c:73]: That's tabs vs. spaces.

Comment (GuuBu) [reply redirect_cache.c:73]: Thanks a lot for your information. Yet I choose not to adjust the formatting this time, thanks

**Pull request merged**
